crates/trie/trie/src/hashed_cursor/post_state.rs

```diff
@@ -25,141 +25,184 @@ where
+    cursor: Option<C>,
+    /// Entry that `database_cursor` is currently pointing to.
+    cursor_entry: Option<(B256, V)>,
+    /// Forward-only in-memory cursor over underlying V.
+    post_state_cursor: ForwardInMemoryCursor<'a, B256, Option<V>>,
+    /// The last hashed key that was returned by the cursor.
     /// De facto, this is a current cursor position.
-    last_account: Option<B256>,
+    last_key: Option<B256>,
+    /// Tracks whether `seek` has been called. Used to prevent re-seeking the DB cursor
+    /// when it has been exhausted by iteration.
+    seeked: bool,
 }
 
-impl<'a, C> HashedPostStateAccountCursor<'a, C>
+impl<'a, C, V> HashedPostStateCursor<'a, C, V>
 where
-    C: HashedCursor<Value = Account>,
+    C: HashedCursor<Value = V>,
+    V: Copy + Clone + std::fmt::Debug,
 {
-    /// Create new instance of [`HashedPostStateAccountCursor`].
-    pub fn new(cursor: C, post_state_accounts: &'a HashedAccountsSorted) -> Self {
-        let post_state_cursor = ForwardInMemoryCursor::new(&post_state_accounts.accounts);
-        let destroyed_accounts = &post_state_accounts.destroyed_accounts;
-        Self { cursor, post_state_cursor, destroyed_accounts, last_account: None }
-    }
-
-    /// Returns `true` if the account has been destroyed.
-    /// This check is used for evicting account keys from the state trie.
+    /// Creates a new post state cursor which combines a DB cursor and in-memory post state updates.
     ///
-    /// This function only checks the post state, not the database, because the latter does not
-    /// store destroyed accounts.
-    fn is_account_cleared(&self, account: &B256) -> bool {
-        self.destroyed_accounts.contains(account)
+    /// # Parameters
+    /// - `cursor`: The database cursor. Pass `None` to indicate:
+    ///   - For accounts: Empty database (no persisted accounts)
+    ///   - For storage: Wiped storage (e.g., via `SELFDESTRUCT` - all previous storage destroyed)
+    /// - `updates`: Pre-sorted post state updates where `Some(value)` indicates an update and
+    ///   `None` indicates a deletion (destroyed account or zero-valued storage slot)
+    pub const fn new(cursor: Option<C>, updates: &'a [(B256, Option<V>)]) -> Self {
+        Self {
+            cursor,
+            cursor_entry: None,
+            post_state_cursor: ForwardInMemoryCursor::new(updates),
+            last_key: None,
+            seeked: false,
+        }
     }
 
-    fn seek_inner(&mut self, key: B256) -> Result<Option<(B256, Account)>, DatabaseError> {
-        // Take the next account from the post state with the key greater than or equal to the
-        // sought key.
-        let post_state_entry = self.post_state_cursor.seek(&key);
+    /// Asserts that the next entry to be returned from the cursor is not previous to the last entry
+    /// returned.
+    fn set_last_key(&mut self, next_entry: &Option<(B256, V)>) {
```

Comment:
This is missing the debug_assert that InMemoryTrieOverlay has for this method

Reply:
See reason from the above comment thread